src/common/com/intellij/plugins/haxe/ide/folding/HaxeFoldingBuilder.java

```diff
@@ -22,37 +23,138 @@
+    final ASTNode openBrace = node.getFirstChildNode();
+    final ASTNode closeBrace = node.getLastChildNode();
+
+    return buildBlockFolding(node, openBrace, closeBrace);
+  }
+
+  private static FoldingDescriptor buildBodyBlockFolding(@NotNull ASTNode node) {
+    final ASTNode openBrace = UsefulPsiTreeUtil.getPrevSiblingSkipWhiteSpacesAndComments(node);
+    final ASTNode closeBrace = UsefulPsiTreeUtil.getNextSiblingSkipWhiteSpacesAndComments(node);
+
+    return buildBlockFolding(node, openBrace, closeBrace);
+  }
+
+  private static FoldingDescriptor buildBlockFolding(@NotNull ASTNode node, ASTNode openBrace, ASTNode closeBrace) {
+    TextRange textRange;
+    if (openBrace != null && closeBrace != null && openBrace.getElementType() == PLCURLY && closeBrace.getElementType() == PRCURLY) {
+      textRange = new TextRange(openBrace.getTextRange().getEndOffset(), closeBrace.getStartOffset());
+    } else {
+      textRange = node.getTextRange();
+    }
+
+    if (textRange.getLength() > PLACEHOLDER_TEXT.length()) {
+      return new FoldingDescriptor(node, textRange);
+    }
+
+    return null;
+  }
+
+  private static FoldingDescriptor buildImportsFolding(@NotNull ASTNode node) {
+    final ASTNode lastImportNode = findLastImportNode(node);
+    if (!node.equals(lastImportNode)) {
+      return new FoldingDescriptor(node, buildImportsFoldingTextRange(node, lastImportNode));
+    }
+    return null;
+  }
+
+  private static TextRange buildImportsFoldingTextRange(ASTNode firstNode, ASTNode lastNode) {
+    ASTNode nodeStartFrom = UsefulPsiTreeUtil.getNextSiblingSkipWhiteSpacesAndComments(firstNode.getFirstChildNode());
+    if (nodeStartFrom == null) {
+      nodeStartFrom = firstNode;
+    }
+    return new TextRange(nodeStartFrom.getStartOffset(), lastNode.getTextRange().getEndOffset());
+  }
+
+  private static boolean isFirstImportStatement(@NotNull ASTNode node) {
+    ASTNode prevNode = node.getTreePrev();
+    while (prevNode != null && prevNode.getElementType() == WHITE_SPACE) {
+      prevNode = prevNode.getTreePrev();
+    }
+    return prevNode == null || !isImportOrUsingStatement(prevNode.getElementType());
+  }
+
+  private static ASTNode findLastImportNode(@NotNull ASTNode node) {
+    ASTNode lastImportNode = node;
+    ASTNode nextNode = lastImportNode.getTreeNext();
+    while (nextNode != null && isApplicableForImportsRegion(nextNode.getElementType())) {
+      lastImportNode = nextNode;
+      nextNode = nextNode.getTreeNext();
+    }
+    if (lastImportNode.getElementType() == WHITE_SPACE) {
+      lastImportNode = lastImportNode.getTreePrev();
+    }
+    return lastImportNode;
+  }
+
+  private static boolean isImportOrUsingStatement(IElementType type) {
+    return type == IMPORT_STATEMENT || type == USING_STATEMENT;
+  }
+
+  private static boolean isApplicableForImportsRegion(IElementType type) {
+    return isImportOrUsingStatement(type) || type == WHITE_SPACE;
```

Comment:
This has a similar problem as above.  There are many other things that are appropriate for an import region.